File: src/email_parser/workflow/nodes/update_memory.py
#!/usr/bin/env python3
"""
Update Memory Node

Final node that:
1. Stores episodic memory for current email
2. Marks email as processed
3. Updates final profile in state
4. Tracks classification changes for dashboard
"""

import logging
from datetime import datetime

from ..state import WorkflowState, get_current_email, add_error
from ...memory.manager import MemoryManager
from ...memory.schemas import build_episodic_memory_id
from ..tracking import WorkflowTracker

logger = logging.getLogger(__name__)


def update_memory_node(
    state: WorkflowState,
    memory_manager: MemoryManager
) -> WorkflowState:
    """
    Update memory store with final results.

    Workflow Steps:
    1. Store episodic memory for this email (evidence trail)
    2. Mark email as processed (for incremental runs)
    3. Update state.updated_profile with latest profile

    Args:
        state: Current workflow state with reconciliation_data
        memory_manager: MemoryManager instance

    Returns:
        Updated state with updated_profile and workflow_completed_at

    Example:
        >>> state = update_memory_node(state, memory_manager)
        >>> # Email marked as processed, memories updated
    """
    try:
        email = get_current_email(state)

        if email is None:
            logger.warning("No email to update in memory")
            return state

        email_id = email.get("id", "unknown")
        email_date = email.get("date", "")
        email_subject = email.get("subject", "")

        logger.info(f"Updating memory for email: {email_id}")

        # Build episodic memory (evidence trail)
        episode_id = build_episodic_memory_id(email_id)

        # Collect taxonomy selections from reconciliation
        taxonomy_selections = []
        confidence_contributions = {}

        for memory in state.get("reconciliation_data", []):
            tax_id = memory.get("taxonomy_id")
            conf = memory.get("confidence", 0.0)

            if tax_id:
                taxonomy_selections.append(tax_id)
                confidence_contributions[tax_id] = conf

        # Create reasoning summary
        reasoning_parts = []
        for memory in state.get("reconciliation_data", []):
            if memory.get("reasoning"):
                reasoning_parts.append(
                    f"{memory.get('value')}: {memory.get('reasoning')}"
                )

        reasoning = "\n".join(reasoning_parts) if reasoning_parts else "Evidence reconciled with existing profile"

        # Store episodic memory
        episodic_data = {
            "email_id": email_id,
            "email_date": email_date,
            "email_subject": email_subject[:200],  # Limit length
            "taxonomy_selections": taxonomy_selections,
            "confidence_contributions": confidence_contributions,
            "reasoning": reasoning,
            "processed_at": datetime.utcnow().isoformat() + "Z",
            "llm_model": "placeholder:phase3-stub"  # Phase 4 will use actual model
        }

        memory_manager.store_episodic_memory(episode_id, episodic_data)
        logger.info(f"Stored episodic memory: {episode_id}")

        # Mark email as processed
        memory_manager.mark_email_as_processed(email_id)
        logger.info(f"Marked email as processed: {email_id}")

        # Update state with latest profile (retrieve all memories)
        all_memories = memory_manager.get_all_semantic_memories()

        # Group by section
        updated_profile = {
            "demographics": [],
            "household": [],
            "interests": [],
            "purchase_intent": [],
            "actual_purchases": []
        }

        for memory in all_memories:
            section = memory.get("section", "unknown")
            if section in updated_profile:
                updated_profile[section].append(memory)

        state["updated_profile"] = updated_profile
        state["workflow_completed_at"] = datetime.utcnow().isoformat() + "Z"

        # Track classification snapshots for dashboard
        # Get or create tracker from state
        tracker = state.get("tracker")
        if tracker is None:
            user_id = state.get("user_id", "unknown")
            logger.warning(f"No tracker found in state, creating new one for user {user_id}")
            tracker = WorkflowTracker(user_id)
            state["tracker"] = tracker

        # Record email processed
        tracker.record_email_processed()

        # Record classification snapshots
        reconciliation_data = state.get("reconciliation_data", [])
        for memory in reconciliation_data:
            taxonomy_id = memory.get("taxonomy_id")
            confidence = memory.get("confidence", 0.0)
            evidence_count = memory.get("evidence_count", 0)

            if taxonomy_id:
                # Check if this is a new classification
                is_new = evidence_count == 1
                tracker.record_classification_change(
                    taxonomy_id=taxonomy_id,
                    confidence=confidence,
                    evidence_count=evidence_count,
                    is_new=is_new
                )

        logger.info(f"Memory update complete")

        return state

    except Exception as e:
        logger.error(f"Error updating memory: {e}", exc_info=True)
        state = add_error(state, f"Memory update failed: {str(e)}")
        return state


def generate_profile_report(state: WorkflowState) -> dict:
    """
    Generate final profile report from updated_profile.

    Args:
        state: Workflow state with updated_profile

    Returns:
        Dictionary with formatted profile report

    Example:
        >>> report = generate_profile_report(state)
        >>> print(f"Demographics: {len(report['demographics'])} classifications")
        >>> print(f"Interests: {len(report['interests'])} classifications")
    """
    updated_profile = state.get("updated_profile", {})

    report = {
        "user_id": state.get("user_id"),
        "generated_at": datetime.utcnow().isoformat() + "Z",
        "demographics": updated_profile.get("demographics", []),
        "household": updated_profile.get("household", []),
        "interests": updated_profile.get("interests", []),
        "purchase_intent": updated_profile.get("purchase_intent", []),
        "actual_purchases": updated_profile.get("actual_purchases", []),
        "metadata": {
            "total_emails_processed": state.get("current_email_index", 0) + 1,
            "workflow_started_at": state.get("workflow_started_at"),
            "workflow_completed_at": state.get("workflow_completed_at"),
            "errors": state.get("errors", []),
            "warnings": state.get("warnings", [])
        }
    }

    return report

File: packages/iab-classifier/src/nodes/updateMemory.ts
/**
 * Update Memory Node
 *
 * 1:1 Port of Python update_memory.py from:
 * src/email_parser/workflow/nodes/update_memory.py (lines 23-159)
 *
 * Final node that:
 * 1. Stores episodic memory for current email
 * 2. Marks email as processed
 * 3. Updates final profile in state
 * 4. Tracks classification changes for dashboard
 */

import { WorkflowState } from '../state'
import { MemoryManager } from '../memory/MemoryManager'

/**
 * Helper Functions
 */

/**
 * Build episodic memory ID
 *
 * Python source: schemas.py:110-125 (build_episodic_memory_id)
 *
 * Format: episodic_email_{email_id}
 */
function buildEpisodicMemoryId(email_id: string): string {
  // Python line 124
  return `episodic_email_${email_id}`
}

/**
 * Get current email being processed
 *
 * Python source: state.py:259-283 (get_current_email)
 */
function getCurrentEmail(state: typeof WorkflowState.State): Record<string, any> | null {
  // Python lines 276-277
  if (!state.emails || state.emails.length === 0) {
    return null
  }

  // Python lines 279-281
  const index = state.current_email_index || 0
  if (index >= state.emails.length) {
    return null
  }

  // Python line 283
  return state.emails[index]
}

/**
 * Update memory node
 *
 * Python source: update_memory.py:23-159 (update_memory_node)
 *
 * @param state Current workflow state with reconciliation_data
 * @param memoryManager MemoryManager instance
 * @returns Updated state with updated_profile and workflow_completed_at
 */
export async function updateMemoryNode(
  state: typeof WorkflowState.State,
  memoryManager: MemoryManager
): Promise<Partial<typeof WorkflowState.State>> {
  try {
    // Python lines 47-51: Get current email
    const email = getCurrentEmail(state)

    if (email === null) {
      console.warn('No email to update in memory')
      return {}
    }

    // Python lines 53-56
    const email_id = email.id || 'unknown'
    const email_date = email.date || ''
    const email_subject = email.subject || ''

    console.info(`Updating memory for email: ${email_id}`)

    // Python lines 59-72: Build episodic memory (evidence trail)
    const episode_id = buildEpisodicMemoryId(email_id)

    // Collect taxonomy selections from reconciliation
    const taxonomy_selections: number[] = []
    const confidence_contributions: Record<number, number> = {}

    for (const memory of (state.reconciliation_data || [])) {
      const tax_id = memory.taxonomy_id
      const conf = memory.confidence || 0.0

      if (tax_id) {
        taxonomy_selections.push(tax_id)
        confidence_contributions[tax_id] = conf
      }
    }

    // Python lines 74-82: Create reasoning summary
    const reasoning_parts: string[] = []
    for (const memory of (state.reconciliation_data || [])) {
      if (memory.reasoning) {
        reasoning_parts.push(
          `${memory.value}: ${memory.reasoning}`
        )
      }
    }

    const reasoning = reasoning_parts.length > 0
      ? reasoning_parts.join('\n')
      : 'Evidence reconciled with existing profile'

    // Python lines 84-96: Store episodic memory
    const episodic_data = {
      email_id,
      email_date,
      email_subject: email_subject.substring(0, 200), // Limit length
      taxonomy_selections,
      confidence_contributions,
      reasoning,
      processed_at: new Date().toISOString(),
      llm_model: 'placeholder:phase3-stub' // Phase 4 will use actual model
    }

    // Python line 96: memory_manager.store_episodic_memory(episode_id, episodic_data)
    await memoryManager.storeEpisodicMemory(episode_id, episodic_data)
    console.info(`Stored episodic memory: ${episode_id}`)

    // Python lines 99-101: Mark email as processed
    // Python line 100: memory_manager.mark_email_as_processed(email_id)
    await memoryManager.markEmailAsProcessed(email_id)
    console.info(`Marked email as processed: ${email_id}`)

    // Python lines 103-119: Update state with latest profile
    // Python line 104: all_memories = memory_manager.get_all_semantic_memories()
    const all_memories = await memoryManager.getAllSemanticMemories()

    // Group by section
    const updated_profile: Record<string, Array<Record<string, any>>> = {
      demographics: [],
      household: [],
      interests: [],
      purchase_intent: [],
      actual_purchases: []
    }

    for (const memory of all_memories) {
      const section = memory.section || 'unknown'
      if (section in updated_profile) {
        updated_profile[section].push(memory)
      }
    }

    // Python lines 120-121
    const workflow_completed_at = new Date().toISOString()

    // Python lines 123-150: Track classification snapshots for dashboard
    // Note: Browser PWA may not need WorkflowTracker (Python-specific)
    // For now, skip tracking implementation

    // TODO: If dashboard tracking is needed, implement:
    // - tracker.recordEmailProcessed()
    // - tracker.recordClassificationChange()

    console.info('Memory update complete')

    // Python line 154: Return updated state
    return {
      updated_profile,
      workflow_completed_at,
    }

  } catch (error) {
    // Python lines 156-159: Error handling
    console.error(`Error updating memory: ${error}`, error)

    const error_message = error instanceof Error ? error.message : String(error)

    return {
      updated_profile: state.existing_profile,
      errors: [...(state.errors || []), `Memory update failed: ${error_message}`],
    }
  }
}

/**
 * Generate profile report
 *
 * Python source: update_memory.py:162-196 (generate_profile_report)
 *
 * @param state Workflow state with updated_profile
 * @returns Dictionary with formatted profile report
 */
export function generateProfileReport(state: typeof WorkflowState.State): Record<string, any> {
  // Python line 177
  const updated_profile = state.updated_profile || {}

  // Python lines 179-194
  const report = {
    user_id: state.user_id,
    generated_at: new Date().toISOString(),
    demographics: updated_profile.demographics || [],
    household: updated_profile.household || [],
    interests: updated_profile.interests || [],
    purchase_intent: updated_profile.purchase_intent || [],
    actual_purchases: updated_profile.actual_purchases || [],
    metadata: {
      total_emails_processed: (state.current_email_index || 0) + 1,
      workflow_started_at: state.workflow_started_at,
      workflow_completed_at: state.workflow_completed_at,
      errors: state.errors || [],
      warnings: state.warnings || [],
    }
  }

  return report
}
